src/feature_store.ts

```diff
@@ -0,0 +1,246 @@
+  private prefix: string;
+  private initedKey: string;
+
+  public constructor(options: LDRedisOptions, logger: ld.LDLogger) {
+    this.state = new BaseRedis(options, logger);
+    this.client = this.state.client;
+    this.logger = logger;
+    this.prefix = this.state.prefix;
+    this.initedKey = this.prefix + "$inited";
+  }
+
+  public getInternal(kind: DataKind, key: string, callback: (item: any) => void): void {
+    this.doGet(kind, key, (item) => {
+      if (item && !item.deleted) {
+        callback(item);
+      } else {
+        callback(null);
+      }
+    });
+  }
+
+  public getAllInternal(kind: DataKind, callback: (items: KeyedItems) => void): void {
+    if (!this.state.connected && !this.state.initialConnect) {
+      this.logger.warn("Attempted to fetch all keys while Redis connection is down");
+      callback(null);
+      return;
+    }
+
+    this.client.hgetall(this.itemsKey(kind), (err, obj) => {
+      if (err) {
+        this.logger.error(`Error fetching "${kind.namespace}" from Redis`, err);
+        callback(null);
+      } else {
+        const results = {};
+        const items = obj;
+        for (const [key, serializedData] of Object.entries(items)) {
+          results[key] = JSON.parse(serializedData);
+        }
+        callback(results);
+      }
+    });
+  }
+
+  public initInternal(allData: FullDataSet, callback: () => void): void {
+    const multi = this.client.multi();
+
+    for (const [kindNamespace, items] of Object.entries(allData)) {
+      const kind = dataKind[kindNamespace];
+      const baseKey = this.itemsKey(kind);
+      const stringified = {};
+      multi.del(baseKey);
+      for (const [key, item] of Object.entries(items)) {
+        stringified[key] = JSON.stringify(item);
+      }
+      // Redis does not allow hmset() with an empty object
+      if (Object.keys(stringified).length > 0) {
+        multi.hmset(baseKey, stringified);
+      }
+    }
+
+    multi.set(this.initedKey, "");
+
+    multi.exec((err) => {
+      if (err) {
+        this.logger.error("Error initializing Redis store", err);
+      }
+      callback();
+    });
+  }
+
+  public upsertInternal(kind: DataKind, item: any, callback: (err: Error, finalItem: any) => void): void {
+    this.updateItemWithVersioning(kind, item, (err, attemptedWrite) => {
+      if (err) {
+        this.logger.error(`Error upserting key ${item.key} in "${kind.namespace}"`, err); // eslint-disable-line quotes
+      }
+      callback(err, attemptedWrite);
+    });
+  }
+
+  public initializedInternal(callback: (result: boolean) => void): void {
+    this.client.exists(this.initedKey, (err, obj) => {
+      callback(!err && !!obj);
+    });
+  }
+
+  public close() {
+    if (this.state.stopClientOnClose) {
+      this.client.quit();
+    }
+  }
+
+  private itemsKey(kind: DataKind): string {
+    return this.prefix + kind.namespace;
+  }
+
+  private doGet(kind: DataKind, key: string, callback: (item: any) => void): void {
```

Comment:
It's generally not a good practice to use the `any` type. What is `item` suppose to be here?

Reply:
It is really an object or null/undefined, but [the interface that we're implementing in this class](https://github.com/launchdarkly/node-server-sdk-private/blob/master/index.d.ts#L567)— which comes from another package that we need to remain compatible with— unfortunately defines it as `any`. I know that we could still use a more specific type in this private helper method, but that callback is coming fro a method in that interface so I think we would need to do a clunky cast in order to use it here.